src/keymap-manager.coffee

```diff
@@ -438,33 +463,41 @@ class KeymapManager
     # state with a timeout.
     if exactMatchCandidates.length > 0
       currentTarget = target
-      while currentTarget? and currentTarget isnt document
+      while not event.handled and currentTarget? and currentTarget isnt document
         exactMatches = @findExactMatches(exactMatchCandidates, currentTarget)
-        for exactMatch in exactMatches
-          if exactMatch.command is 'native!'
-            @clearQueuedKeystrokes()
-            return
+        for exactMatchCandidate in exactMatches
+          if exactMatchCandidate.command is 'native!'
+            return {keystroke, queuedKeystrokes: null}
 
-          if exactMatch.command is 'abort!'
-            @clearQueuedKeystrokes()
+          if exactMatchCandidate.command is 'abort!'
             event.preventDefault()
-            return
+            return {keystroke, queuedKeystrokes: null}
 
-          if exactMatch.command is 'unset!'
+          if exactMatchCandidate.command is 'unset!'
             break
 
-          foundMatch = true
-          break if partialMatches.length > 0
-          @clearQueuedKeystrokes()
-          @cancelPendingState()
-          if @dispatchCommandEvent(exactMatch.command, target, event)
+          exactMatch = exactMatchCandidate
+          if partialMatches.length > 0
+            # When there is a set of bindings like `'ctrl-y', 'ctrl-y ^ctrl'`,
+            # and a `ctrl-y` comes in, this will allow the `ctrl-y` command to be
+            # dispatched without waiting for any other keystrokes
+            allPartialMatchesContainKeyupRemainder = true
+            for partialMatch in partialMatches
+              if keydownExactMatchCandidates.indexOf(partialMatch) < 0
+                allPartialMatchesContainKeyupRemainder = false
+            break if allPartialMatchesContainKeyupRemainder is false
+          else
+            queuedKeystrokes = null
+
+          if @dispatchCommandEvent(exactMatchCandidate.command, target, event)
             @emitter.emit 'did-match-binding', {
               keystrokes,
               eventType: event.type,
-              binding: exactMatch,
+              binding: exactMatchCandidate,
               keyboardEventTarget: target
             }
-            return
+            event.handled = true # Kicks it out of the parent loop
```

Comment:
We set this on the event itself for the sake of the replay. These events shouldn't generate another exact match command dispatch.